design/ceph/ceph-orchestrator-osd-add-remove.md

```diff
@@ -0,0 +1,266 @@
+nodes in the Kubernetes cluster, to a specific node, or via disk pattern matching (in two different
+ways). To avoid the complexity of having to manage the logic for determining how to remove a node
+from the `CephCluster` resource, OSD removal will **not** alter the `CephCluster` resource's
+`storage` specification with the exception of Drive Groups when needed. While modifying these items
+might be convenient for users, there are a lot of opportunities in development to miss corner cases
+or introduce logic bugs, which we should avoid.
+
+Users should be advised in documentation that if they plan to use the Ceph CLI or Dashboard to
+manage OSD addition in the Rook-Ceph cluster, it is recommended to set `useAllDevices: false`, leave
+both disk pattern matching empty, and to not specify any other disks in the `CephCluster` resource
+to avoid giving Rook the ability to re-add OSDs when the user does not wish it to. We should also
+note to users that Rook can automatically re-add removed OSDs if they are zapped manually by the
+user via documentation to avoid possible confusion and annoyance.
+
+#### Zapping OSDs
+This design proposes to add a section to the `CephCluster` resource's `storage` configuration
+section to allow OSDs to be zapped by Rook. Note that zapping is not the same as removing an OSD.
+Users are still expected to use either manual steps or the Ceph Orchestrator interfaces to remove
+OSDs from Ceph.
+
+```yaml
+apiVersion: ceph.rook.io/v1
+kind: CephCluster
+spec:
+  storage: # preexisting
+    zapOSDs: # proposed new
+      - FSID: < string, FSID of OSD to zap > # proposed new
+        controlledByCephOrchestrator: < bool, default false >
+```
+
+Zapping disks requires running a `ceph-volume` command to remove the disk. Ceph's Rook Orchestrator
+Mgr Module can modify the `CephCluster` resource to identify a specific OSD which Rook should zap
+using `ceph-volume` by adding the OSD to the `storage:zapOSDs` spec.
+
+Rook matches OSDs to remove by the OSD `FSID`, which will be unique for every OSD even if an OSD
+with the same ID is recreated. This will prevent Rook from accidentally removing new OSDs in the
+future if recreated on the same disk or with the same OSD ID.
+
+In the OSD provisioning Job, Rook will check if the OSD with the corresponding FSID exists, and then
+Rook will zap the OSD using `ceph-volume lvm zap --destroy --osd-fsid <FSID>`.
+
+The Rook Orchestrator Mgr Module should set the `controlledByCephOrchestrator` field `true` to give
+it permission to edit and remove the item in the `zapOSDs` list. Users can be made aware that if
+they set this value `true`, the Orchestrator Module may modify or delete that item automatically.
+
+When the Orchestrator Module specifies an OSD to remove in the `CephCluster` resource, the Module
+should also edit the Drive Group(s) add the corresponding disk to the blacklist section of the Drive
```

Comment:
If no Drive Groups are used` ( a cluster created with  useAllNodes: true,  useAllDevices: true).`  How the user can prevent the automatic recreation of the deleted  OSD ( because i think that the blacklist is only into the drive groups).

Reply:
This is noted below, and preventing re-addition by other (non-Drive-Group) methods is covered earlier in the doc.